aim/sdk/index_manager.py

```diff
@@ -0,0 +1,180 @@
+        ):
+            pass
+
+    def _run_forever(self):
+        lock_path = self.locks_dir / 'run_index'
+        logger.debug(f'Trying to acquire indexing lock for repo \'{self.repo.path}\'...')
+        lock = FileLock(str(lock_path))
+        lock.acquire(timeout=-1)  # block till lock can be acquired
+        logger.debug('Lock acquired! Running...')
+        idle_cycles = 0
+        while True:
+            self._indexing_in_progress = False
+            for run in self._next_unindexed_run():
+                logger.info(f'Found un-indexed run {run.hash}. Indexing...')
+                self._indexing_in_progress = True
+                idle_cycles = 0
+                run.finalize()
+                # sleep for 2 seconds to release index db lock in between and allow
+                # potential running jobs to properly finalize and index Run.
+                sleep_interval = 2
+                time.sleep(sleep_interval)
+            if not self._indexing_in_progress:
+                idle_cycles += 1
+                sleep_interval = 2 * idle_cycles if idle_cycles < 5 else 10
+                logger.info(f'No un-indexed runs found. Next check will run in {sleep_interval} seconds. '
+                            f'Waiting for un-indexed run...')
+                time.sleep(sleep_interval)
+
+    def _run_has_progress_file(self, run_hash: str) -> bool:
+        # This method is required to detect the case when Run was finalized after collecting
+        # un-indexed runs. Such case may occur since there's a significant interval between collecting
+        # runs and actual indexing and Run might be finalized manually during that period.
+        run_progress = os.path.join(self.repo.path, 'meta', 'progress', run_hash)
+        return os.path.exists(run_progress)
+
+    def _is_run_in_progress(self, run_hash: str) -> bool:
+        lock_path = os.path.join(self.repo.path, 'meta', 'locks', run_hash)
+        try:
+            fl = FileLock(lock_path, timeout=0)
+            fl.acquire()
+        except TimeoutError:
+            return True
+        else:
+            fl.release()
+            return False
+
+    def _unindexed_runs(self) -> List[str]:
+        progress_dir = os.path.join(self.repo.path, 'meta', 'progress')
+        runs_with_progress = os.listdir(progress_dir)
+        run_hashes = sorted(filter(lambda r: not self._is_run_in_progress(r), runs_with_progress),
+                            key=lambda r: os.path.getmtime(os.path.join(progress_dir, r)))
+        return run_hashes
+
+    def _in_progress_runs(self) -> List[str]:
+        progress_dir = os.path.join(self.repo.path, 'meta', 'progress')
+        runs_with_progress = os.listdir(progress_dir)
+        run_hashes = sorted(filter(lambda r: self._is_run_in_progress(r), runs_with_progress),
+                            key=lambda r: os.path.getmtime(os.path.join(progress_dir, r)))
+        return run_hashes
+
+    def _run(self, run_hash):
```

Comment:
what's the point wrapping the `Run` inside a private method?

Reply:
just to make the line shorter in a couple of places 😄
`self._run(run_hash)`
vs.
`Run(run_hash, repo=self.repo, system_tracking_interval=None)`

Reply:
okay, now seeing it side by side makes sense 😆